File: playwright-csharp/src/Runtime/ApiSession.cs
using Microsoft.Playwright;
using TestTracks.Playwright.CSharp.Configuration;

namespace TestTracks.Playwright.CSharp.Runtime;

/// <summary>
/// Wraps a Playwright API request context and the Playwright instance that owns it.
/// </summary>
public sealed class ApiSession : IAsyncDisposable
{
    private ApiSession(IPlaywright playwright, IAPIRequestContext request)
    {
        Playwright = playwright;
        Request = request;
    }

    public IPlaywright Playwright { get; }

    public IAPIRequestContext Request { get; }

    /// <summary>
    /// Creates an API request context with a shared base URL and JSON-friendly default headers.
    /// </summary>
    public static async Task<ApiSession> CreateAsync(Uri baseUrl)
    {
        var playwright = await Microsoft.Playwright.Playwright.CreateAsync();
        var request = await playwright.APIRequest.NewContextAsync(new()
        {
            BaseURL = baseUrl.ToString(),
            ExtraHTTPHeaders = new Dictionary<string, string>
            {
                ["Accept"] = "application/json"
            }
        });

        return new ApiSession(playwright, request);
    }

    /// <inheritdoc />
    public async ValueTask DisposeAsync()
    {
        var failures = new List<Exception>();

        try
        {
            await Request.DisposeAsync();
        }
        catch (Exception ex)
        {
            failures.Add(new InvalidOperationException("API request context disposal failed.", ex));
        }

        try
        {
            Playwright.Dispose();
        }
        catch (Exception ex)
        {
            failures.Add(new InvalidOperationException("API Playwright disposal failed.", ex));
        }

        if (failures.Count > 0)
        {
            throw new AggregateException(failures);
        }
    }
}

File: playwright-java/src/main/java/testtracks/runtime/ApiSession.java
package testtracks.runtime;

import com.microsoft.playwright.APIRequest;
import com.microsoft.playwright.APIRequestContext;
import com.microsoft.playwright.Playwright;
import java.net.URI;
import java.util.ArrayList;
import java.util.Map;

public final class ApiSession implements AutoCloseable {
  private final Playwright playwright;
  private final APIRequestContext request;

  private ApiSession(Playwright playwright, APIRequestContext request) {
    this.playwright = playwright;
    this.request = request;
  }

  public APIRequestContext request() {
    return request;
  }

  public static ApiSession create(URI baseUrl) {
    var playwright = Playwright.create();
    var request =
        playwright
            .request()
            .newContext(
                new APIRequest.NewContextOptions()
                    .setBaseURL(baseUrl.toString())
                    .setExtraHTTPHeaders(Map.of("Accept", "application/json")));

    return new ApiSession(playwright, request);
  }

  @Override
  public void close() {
    var failures = new ArrayList<Exception>();

    try {
      request.dispose();
    } catch (Exception ex) {
      failures.add(new IllegalStateException("API request context disposal failed.", ex));
    }

    try {
      playwright.close();
    } catch (Exception ex) {
      failures.add(new IllegalStateException("API Playwright disposal failed.", ex));
    }

    if (!failures.isEmpty()) {
      throw new IllegalStateException("API session disposal failed.", failures.getFirst());
    }
  }
}
